src/task_manager.rs

```diff
@@ -0,0 +1,335 @@
+                        RecipeStep::GetCompletionsForDefId => {
+                            if let Ok(def_id) = argument.downcast::<DefId>() {
+                                self.send_to_type_checker.as_ref().map(|x| {
+                                    x.send(MsgFromManager::Message(
+                                        TypeMessage::CompletionsForDefId(task_id, *def_id),
+                                    ))
+                                    .unwrap()
+                                });
+                            } else {
+                                unimplemented!("Internal error: malformed GetCompletionsForDefId");
+                            }
+                        }
+                        RecipeStep::RespondWithType => {
+                            if let Ok(ty) = argument.downcast::<String>() {
+                                self.send_to_lsp_responder.as_ref().map(|x| {
+                                    x.send(MsgFromManager::Message(LspResponse::Type(task_id, *ty)))
+                                        .unwrap()
+                                });
+                            } else {
+                                unimplemented!("Internal error: malformed RespondWithCompletion");
+                            }
+                        }
+                        RecipeStep::RespondWithCompletions => {
+                            if let Ok(completions) = argument.downcast::<Vec<(String, String)>>() {
+                                self.send_to_lsp_responder.as_ref().map(|x| {
+                                    x.send(MsgFromManager::Message(LspResponse::Completions(
+                                        task_id,
+                                        *completions,
+                                    )))
+                                    .unwrap()
+                                });
+                            } else {
+                                unimplemented!("Internal error: malformed RespondWithCompletion");
+                            }
+                        }
+                        RecipeStep::RespondWithInitialized => {
+                            self.send_to_lsp_responder.as_ref().map(|x| {
+                                x.send(MsgFromManager::Message(LspResponse::Initialized(task_id)))
+                                    .unwrap()
+                            });
+                        }
+                    }
+                }
+            }
+            None => {
+                //Do nothing as task has completed or it has been cancelled
+            }
+        }
+    }
+
+    fn do_recipe_for_lsp_request(&mut self, lsp_request: LspRequest) {
+        match lsp_request {
+            LspRequest::TypeForPos(task_id, position) => {
+                let recipe = vec![
+                    RecipeStep::GetDefIdForPosition,
+                    RecipeStep::GetTypeForDefId,
+                    RecipeStep::RespondWithType,
+                ];
```

Comment:
This sort of "IR layer" is interesting; I had expected to encode this with Rust closures (more similar to futures), but I could imagine that having more explicit steps would pay off.